### PR TITLE
Add task to generate upm packages and to publish them

### PR DESCRIPTION
**Pull request by Azurelol**
## Description

Adds a conventional task to generate the UPM package for the given unity project. This involves adding the JFrog artifactory plugin, and hooking into the `publish` task.

## Changes
* ![ADD] `upmPack` plugin task
* ![ADD] `artifactory` plugin from Jfrog

[NEW]:https://atlas-resources.wooga.com/icons/icon_new.svg "New"
[ADD]:http://resources.atlas.wooga.com/icons/icon_add.svg "Add"
[IMPROVE]:http://resources.atlas.wooga.com/icons/icon_improve.svg "IMPROVE"
[CHANGE]:http://resources.atlas.wooga.com/icons/icon_change.svg "Change"
[FIX]:http://resources.atlas.wooga.com/icons/icon_fix.svg "Fix"
[UPDATE]:http://resources.atlas.wooga.com/icons/icon_update.svg "Update"

[BREAK]:http://resources.atlas.wooga.com/icons/icon_break.svg "Remove"
[REMOVE]:http://resources.atlas.wooga.com/icons/icon_remove.svg "Remove"
[IOS]:http://resources.atlas.wooga.com/icons/icon_iOS.svg "iOS"
[ANDROID]:http://resources.atlas.wooga.com/icons/icon_android.svg "Android"
[WEBGL]:http://resources.atlas.wooga.com/icons/icon_webGL.svg "Web:GL"

### CODE REVIEW

File: src/integrationTest/groovy/wooga/gradle/wdk/unity/PublishWdkUnityIntegrationSpec.groovy [hunk 1]
```diff
@@ -89,7 +89,7 @@ class PublishWdkUnityIntegrationSpec extends com.wooga.gradle.test.IntegrationSp
         assert packageDir.exists()
 
         and: "configuration of the main plugin and the artifactory plugin"
-        String nugetKey = System.getenv("NUGET_KEY")
+        String nugetKey = System.getenv("repositoryCredentials") ?: System.getenv("NUGET_KEY")
```
Comment (Larusso) [inline src/integrationTest/groovy/wooga/gradle/wdk/unity/PublishWdkUnityIntegrationSpec.groovy:92]: We should not use NUGET_KEY at all. Or bind it to a scoped credential in the Jenkinsfile / locally.

**Review by Larusso — CHANGES_REQUESTED** [review]

File: build.gradle [hunk 2]
```diff
@@ -22,6 +22,11 @@ plugins {
     id "net.wooga.cve-dependency-resolution" version "0.4.0"
 }
 
+repositories  {
+    mavenLocal()
```
Comment (Larusso) [inline build.gradle:26]: please remove mavenLocal here

File: src/integrationTest/groovy/wooga/gradle/wdk/unity/WdkUnityIntegrationSpec.groovy [hunk 3]
```diff
@@ -129,6 +131,8 @@ class WdkUnityIntegrationSpec extends UnityIntegrationSpec {
         WdkUnityPlugin.SONARQUBE_BUILD_TASK_NAME                | "test"                                                  | ["test", "sonarBuildWDK"]
         WdkUnityPlugin.SONARQUBE_TASK_NAME                      | "test"                                                  | ["test", "sonarqube"]
         WdkUnityPlugin.SONARQUBE_TASK_NAME                      | WdkUnityPlugin.SONARQUBE_BUILD_TASK_NAME                | ["sonarBuildWDK", "sonarqube"]
+        // TODO: Won't work unless it's properly set, sadly
+        //WdkUnityPlugin.GENERATE_UPM_PACKAGE_TASK_NAME           | WdkUnityPlugin.GENERATE_META_FILES_TASK_NAME            | ["upmPack"]
```
Comment (Larusso) [inline src/integrationTest/groovy/wooga/gradle/wdk/unity/WdkUnityIntegrationSpec.groovy:135]: I don't get what the comment wants to convey here.

File: src/main/groovy/wooga/gradle/wdk/unity/WdkUnityPlugin.groovy [hunk 4]
```diff
@@ -47,24 +57,30 @@ import wooga.gradle.wdk.unity.actions.IOSResourceCopyAction
...
-    static String GROUP = "wdk"
-    static String EXTENSION_NAME = "wdk"
+
     static String ANDROID_RESOURCES_CONFIGURATION_NAME = "android"
     static String IOS_RESOURCES_CONFIGURATION_NAME = "ios"
     static String WEBGL_RESOURCES_CONFIGURATION_NAME = "webgl"
     static String RUNTIME_CONFIGURATION_NAME = "runtime"
 
+    static String GENERATE_UPM_PACKAGE_TASK_NAME = "upmPack"
+    static String GENERATE_META_FILES_TASK_NAME = "generateMetaFiles"
```
Comment (Larusso) [inline src/main/groovy/wooga/gradle/wdk/unity/WdkUnityPlugin.groovy:82]: I would not yet put this as a public task to be fetchable via constants.

File: src/main/groovy/wooga/gradle/wdk/unity/WdkUnityPlugin.groovy [hunk 5]
```diff
@@ -114,6 +131,25 @@ class WdkUnityPlugin implements Plugin<Project> {
...
+                if (files != null &&  files.length == 1 && files[0].isDirectory()) {
+                    def packageDir = files[0]
+                    result = woogaDir.dir(packageDir.name)
+                }
+            }
+            if (result == null) {
+                logger.warn("Could not deduce the package directory for this wdk")
+            }
+            result
+        }))
```
Comment (Larusso) [inline src/main/groovy/wooga/gradle/wdk/unity/WdkUnityPlugin.groovy:152]: Why doing it this way around? You could also find all `package.json` files in the Assets directory and generate pack tasks for each of them. Similar to how we did it in paket. I think this method as is would create more harm. I would rather try to find the package.json (the first one as a start) or let this variable be configuable only. The whole system falls flat when a .meta file lands in the directory by the way.

**Review by Larusso — CHANGES_REQUESTED** [review]

File: build.gradle (hunk 2 above)
Comment (Azurelol) [reply build.gradle:26]: Will do

File: src/main/groovy/wooga/gradle/wdk/unity/WdkUnityPlugin.groovy (hunk 4 above)
Comment (Azurelol) [reply src/main/groovy/wooga/gradle/wdk/unity/WdkUnityPlugin.groovy:82]: I will make it private then.

File: src/integrationTest/groovy/wooga/gradle/wdk/unity/PublishWdkUnityIntegrationSpec.groovy (hunk 1 above)
Comment (Azurelol) [reply src/integrationTest/groovy/wooga/gradle/wdk/unity/PublishWdkUnityIntegrationSpec.groovy:92]: Alright. I had set it this way for local testing.

File: src/main/groovy/wooga/gradle/wdk/unity/WdkUnityPlugin.groovy (hunk 5 above)
Comment (Azurelol) [reply src/main/groovy/wooga/gradle/wdk/unity/WdkUnityPlugin.groovy]: I will iterate on the test I wrote for this to check the other cases.

**Review by Larusso — APPROVED** [review]

**Pull request merged**